http/peer-id-auth.md

```diff
@@ -0,0 +1,194 @@
+| public-key       | A base64 encoded value of peer's public key. The key itself is encoded per the [Peer ID spec].                                                                                                                                           |
+| opaque           | An base64 encoded opaque to the client blob generated by the server. If a client receives this it must return it. A server may use this to authenticate statelessly. For example, it could store the challenge-client and a expiry time. |
+
+Params are encoded per [RFC 9110 auth-param's ABNF](https://datatracker.ietf.org/doc/html/rfc9110#name-collected-abnf). Generally it'll be something like: `hostname="example.com", challenge-server=base64EncodedVal`
+
+## Signing
+
+Signatures sign some set of parameters. The parameters are sorted
+alphabetically, prepended with a varint length prefix, and concatenated together
+to form the data to be signed. The signing algorithm is defined by the key type
+used. Refer to the [Peer ID
+spec] for
+specifics on the signing algorithm. The set of parameters is prefixed with the auth scheme "libp2p-PeerID"
+
+As an example, if we wanted to sign the parameters `hostname="example.com",
+challenge-client=base64String` we would first structure the parameters as a byte
+slice containing:
+```
+libp2p-PeerID<varintprefix>challenge-client=<base64String><varintprefix>hostname="example.com"
+```
+
+Then sign the resulting byte slice. See the test vectors below for a
+examples.
+
+
+## Base64 Encoding
+
+The base64 encoding follows Base 64 Encoding with URL and Filename Safe Alphabet
+from [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648#section-5). The
+reason this is not a multibase is to aid clients or servers who can not or
+prefer not to import a multibase dependency.
+
+## Mutual Client and Server Peer ID
+
+The following protocol allows both the client and server to authenticate each
+other's peer id by having them each sign a challenge issued by the other. The
+protocol operates as follows:
+
+1. The client makes a POST request to the authentication URI.
+2. The server responds with status code 401 (Unauthorized) and set the header:
+   ```
+   WWW-Authenticate: libp2p-PeerID challenge-client=<base64-encoded-challenge>, opaque=<base64-encoded-opaque-value>
+   ```
+   The opaque parameter is opaque to client. The client MUST return the opaque
+   parameter back to the server. The server MAY use the opaque parameter to
+   encode state.
+3. The client makes another POST request to the authentication URI and sets the header:
+   ```
+   Authorization: libp2p-PeerID peer-id=<string-representation-of-client-peer-id>, opaque=<opaque-from-server>, challenge-server=<base64-encoded-challenge-server>, sig=<base64-signature-bytes>
+   ```
+
+   The `sig` param represents a signature over the parameters:
```

Comment:
It would be nice to include the certificate (or public key?) hash, if possible. That way, _if_ the client has some way of validating it they can not rely on the CA system and instead validate the certificate.

Reply:
This could be easy to do, but I'm not sure it would be used. The optimal solution here is to authenticate with some TLS exported key material ([RFC 5705](https://datatracker.ietf.org/doc/html/rfc5705)). That would bind the peer id to the TLS session and can happen with no additional round trips. If a peer can access the certificate they likely can access the key material exporter.

The only problem with this approach is that the key material exporters aren't yet exposed by browsers. But I expect that to change soon-ish as https://datatracker.ietf.org/doc/draft-ietf-httpbis-unprompted-auth/ gets published and implemented by browsers.

My thought here is to focus on the browser use case, and, if it would be used and useful, create a new spec for the exported key material scenario.

Reply:
> The optimal solution here is to authenticate with some TLS exported key material ([RFC 5705](https://datatracker.ietf.org/doc/html/rfc5705)). That would bind the peer id to the TLS session and can happen with no additional round trips. If a peer can access the certificate they likely can access the key material exporter.

I agree, but that's likely harder to get at.

> My thought here is to focus on the browser use case, and, if it would be used and useful, create a new spec for the exported key material scenario.

What about making the format extensible? I.e., allow the server to sign multiple things that can be extended later?

Reply:
> I agree, but that's likely harder to get at.

Ah, I see. That's what that proposal is about. Nice!

Reply:
> What about making the format extensible? I.e., allow the server to sign multiple things that can be extended later?

There is room in this spec to extend this later. A future version could for example define a new parameter that is passed by the client. Example: a SessionID parameter which is derived from keying material from the tls session just for this purpose, the server would check that it matches the expected SessionID it sees or fail the request.

Reply:
This has to be covered by a signature. I.e., either:

1. It's covered by the client signature and the server refuses to authenticate if it doesn't match
2. It's covered by the server's signature and the client rejects the authentication if it doesn't match.

I want to make sure there's some sane upgrade path where we can add support for this later.

Reply:
Yes, sorry. The signature SHOULD include the SessionID as part of the parameters it signs.

Reply:
Just to be clear with the example for a future extension here:

1. Each side derives a `TLSSessionID` using TLS' Keying Material exporters in a TBD way.
2. Each side passes that as a parameter to the `libp2p-PeerID` auth scheme. And includes it as parameter in the signature.
3. Each side MUST verify the provided `TLSSessionID` matches the expected value (as well as verifying the signature as before).
4. Either side MUST fail the authentication if there is a mismatch.

Both sides need to sign and compare it.